panos_upgrade_assurance/utils.py

```diff
@@ -187,80 +188,136 @@ def __init__(
-            return True
-        else:
+
+        if f"!{element}" in config:
             return False
+        elif ( element in config or "all" in config ):
+            return True
+        elif all_exclusive_check and ConfigParser.is_all_exclusive(config):
+            return True
+
+        return False
 
     @staticmethod
-    def _extract_element_name(config: Union[str, dict]) -> str:
+    def is_element_explicit_excluded(element, config):
+        """Method verifying if a given element should be excluded.
+
+        TODO docstring
+        This method is required for snapshot comparison when we need to check if we should further compare
+        nested dicts - it doesnot work to check with is_element_included for that case since nested dict key
+        might not be included but nested keys might be subject to comparison
+        """
+        if config and f"!{element}" in config:  # if config isnot none/empty and !element is in config
+            return True
+
+        return False
+
+    @staticmethod
+    def _extract_element_name(config_element: Union[str, dict]) -> str:
         """Method that extracts the name from a config element.
 
         If a config element is a string, the actual config element is returned. For elements of a dictionary type, the
         1<sup>st</sup> key is returned.
 
         # Parameters
 
-        config (str, dict): A config element to provide a name for.
+        config_element (str, dict): A config element to provide a name for.
 
         # Raises
 
-        WrongDataTypeException: Thrown when config does not meet requirements.
+        WrongDataTypeException: Thrown when config_element does not meet requirements.
 
         # Returns
 
         str: The config element name.
 
         """
-        if isinstance(config, str):
-            return config
-        elif isinstance(config, dict):
-            if len(config) == 1:
-                return list(config.keys())[0]
+        if isinstance(config_element, str):
+            return config_element
+        elif isinstance(config_element, dict):
+            if len(config_element) == 1:
+                return list(config_element.keys())[0]
             else:
                 raise exceptions.WrongDataTypeException(
                     "Dict provided as config definition has incorrect format, it is supposed to have only one key {key:[]}"
                 )
         else:
             raise exceptions.WrongDataTypeException("Config definition is neither string or dict")
 
-    def _expand_all(self) -> None:
-        """Expand key word `'all'` to  `self.valid_elements`.
+    def _iter_extracted_config_names(self, config):
+        """
+        TODO docstring
+        """
+        for config_name in config:
+            yield ConfigParser._extract_element_name(config_name)
+
+    def _get_pure_element_name(self, element):
+        return element[1:] if element.startswith("!") else element
+
+    def _get_config_element_by_name(self, element):
```

Comment:
Test with:
```
- nics:
```

Reply:
This already has a check in `run_readiness_checks` and `compare_snapshots` which check keys with None value and set to empty dict.
And this is not a supported config for snapshots, it only accepts str values.